aten/src/ATen/native/Integration.cpp

```diff
@@ -0,0 +1,72 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/core/DimVector.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+namespace native {
+namespace {
+
+// The estimated integral of a function y of x,
+// sampled at points (y_1, ..., y_n) that are separated by distance (dx_1, ..., dx_{n-1}),
+// is given by the trapezoid rule:
+//
+// \sum_{i=1}^{n-1}  dx_i * (y_i + y_{i+1}) / 2
+//
+// TODO: if we extend TensorIterator to accept 3 inputs,
+// we can probably make this a bit more performant.
+static Tensor do_trapz(const Tensor& y, const Tensor& dx, int64_t dim) {
+    Tensor left = y.slice(dim, 0, -1);
+    Tensor right = y.slice(dim, 1);
+
+    return ((left + right) * dx).sum(dim) / 2.;
+}
+
+// When dx is constant, the above formula simplifies
+// to dx * [(\sum_{i=1}^n y_i) - (y_1 + y_n)/2]
+static Tensor do_trapz(const Tensor& y, double dx, int64_t dim) {
+    return (y.sum(dim) - (y.select(dim, 0) + y.select(dim, -1)) * (0.5)) * dx;
+}
+
+static Tensor zeros_like_except(const Tensor& y, int64_t dim) {
+    auto sizes = y.sizes().vec();
+    dim = maybe_wrap_dim(dim, y.dim());
+    sizes.erase(sizes.begin() + dim);
+    return at::zeros(sizes, y.options());
+}
+
+}
+
+Tensor trapz(const Tensor& y, const Tensor& x, int64_t dim) {
```

Comment:
Does `x` have to be sorted in ascending order?
```
>>> np.trapz([0, 2, 1], x=[0, 2, 1])
0.5
```
This seems like it should be a user error to me

Reply:
In this diff, we do the same thing as numpy:

```
>>> torch.trapz(torch.tensor([0,2,1],dtype=float),x=torch.tensor([0,2,1],dtype=float))
tensor(0.5000, dtype=torch.float64)
```

It sort of makes sense, if you think of this as estimating \int_{x=0}^2 x\, dx + \int_{x=2}^1 x\, dx

But it does seem unusual at least to want your integration variable to be non-monotonic like that... do you think we should keep the Numpy behavior or just error out if `x` isn't monotonic?

Reply:
It seems like this is a "feature": https://github.com/numpy/numpy/issues/4601 though I can't see the discussion they linked

Reply:
yeah OK, thanks for digging that up. If the behavior is intended then I'll have us match it and just document it.